### ISSUE TITLE
spirv-reduce and undefined behaviour

### ISSUE THREAD

**Issue by s-perron** [KhronosGroup/SPIRV-Tools#3760]
I've been using spirv-reduce a lot more recently.  There is one feature I would like added.  When dealing with compile time failures like a segfault in spirv-opt or spirv-opt generating invalid spir-v, the current set of passes is good.  However, I'm now trying to reduce spir-v where the error shows up at run time.  The typical test will be:

1. Run the test with a known good driver/compiler to produce an image
1. Run the test with a bad driver/compiler to produce and an image.
1. Compare the two images, report a failure if they are different.

When running the spirv-reduce, it does things like remove store to the outputs or replaces values with undefs.  When that happens, the reduced spirv may no longer demonstrate the original bug.  The images are different because the random undefined values are different.

Is it possible to add an option that says "do not introduce undefined values or behaviour".  Then we can remove passes or adjust passes based on that option.

**Comment by s-perron** [comment]
@afd What do you think?

**Comment by afd** [comment]
Good to hear that you've been using spirv-reduce.

Adding an option sounds like a good idea, and it should be easy to make the tool avoid some of the straightforward things that introduce undefined behaviour, like uses of OpUndef or explicitly removing stores to output variables.

More generally, totally eliminating undefined behaviour is very hard, though: any change to the module risks leading to a division by zero, or changing some computation that has a data dependence on an access chain index, leading to an OOB access, etc.

So, let's do the easy stuff and then see how well that works for your use cases.

@kuhar had a proposal to build [spirv-lint](https://github.com/KhronosGroup/SPIRV-Tools/issues/3196) - if we had a tool like that, it could be part of the interestingness test that spirv-reduce uses - i.e., a shader is regarded as uninteresting if it doesn't pass the linter.  This is how C-Reduce - reducer of C programs works - you write an interestingness test that uses a number of static analyses (or even lightweight dynamic analyses) to say that a program is uninteresting if it looks like it triggers UB.

**Comment by afd** [comment]
@s-perron Do you have an example (even if not shareable publicly) you could use to give feedback on this?

I was thinking, for starters, of having an option, --avoid-ub, that will:

- Disable reduction passes that increase use of OpUndef
- Ensure that a store to an output variable is never removed (but the value being stored might be changed to something simpler)
- Add initializers to Function and Private variables that don't already have them
- [Possibly] Never remove a store to a Function or Private variable that is uninitialized at its declaration (I fear that this might inhibit reduction quite a bit)

It would be interesting to know if these alone would be useful, but I don't have a good example to test it out on.

From your experience using the tool recently, do those ideas seem that they might help?

And by "output variable" I was thinking of the Output storage class, but let me know if other storage classes are important for your use cases, e.g. SSBOs.

**Comment by s-perron** [comment]
Yes that should be useful.  I don't have any examples on hand.  This is similar to the way I hacked spirv-reduce so I could make progress.  However, after reducing, I deleted the setup I had.  They existed on a vm that has been deleted.

> [Possibly] Never remove a store to a Function or Private variable that is uninitialized at its declaration (I fear that this might inhibit reduction quite a bit)

In my hack, I turned off the removal of all OpStores.  It did not have a negative effect because, I could usually run spirv-opt before running spirv-reduce.  Spirv-opt would remove all stores to function and private scope variable by propagating the values to the loads.  I seem to remember someone mentioning that you might want to run dead-code-elimination from time to time while reducing.  That could make this change possible.
